buku

```diff
@@ -5778,11 +5772,14 @@ def parse_range(tokens: Optional[str | Sequence[str] | Set[str]],  # Optional[st
     for token in ([tokens] if isinstance(tokens, str) else tokens):
         for idx in token.split(','):
             if is_int(idx):
-                result |= {int(idx)}
+                result |= ({int(idx)} if not idx.startswith('-') or maxidx is None else
+                           set(range(maxidx, max(0, maxidx + int(idx)), -1)))
             elif '-' in idx:
                 l, r = map(int, idx.split('-'))
                 if l > r:
                     l, r = r, l
+                if maxidx is not None:
+                    r = min(r, maxidx)
```

Comment:
Limiting the upper end of the range when the limit is provided
(lower end is naturally `0` since negative numbers cannot be used with the range descriptors syntax)